cmd/codegen/common.go

```diff
@@ -0,0 +1,59 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"dagger.io/dagger"
+	"github.com/dagger/dagger/cmd/codegen/generator"
+)
+
+var (
+	outputDir             string
+	lang                  string
+	introspectionJSONPath string
+	bundle                bool
+)
+
+func relativeTo(basepath string, tarpath string) (string, error) {
+	basepath, err := filepath.Abs(basepath)
+	if err != nil {
+		return "", err
+	}
+	tarpath, err = filepath.Abs(tarpath)
+	if err != nil {
+		return "", err
+	}
+	return filepath.Rel(basepath, tarpath)
+}
+
+func getGlobalConfig(ctx context.Context) (generator.Config, error) {
+	cfg := generator.Config{
+		Lang:      generator.SDKLang(lang),
+		OutputDir: outputDir,
+		Bundle:    bundle,
+	}
+
+	// If a module source ID is provided or no introspection JSON is provided, we will query
+	// the engine so we can create a connection here.
+	if moduleSourceID != "" || introspectionJSONPath == "" {
+		dag, err := dagger.Connect(ctx)
+		if err != nil {
+			return generator.Config{}, fmt.Errorf("failed to connect to engine: %w", err)
+		}
+
+		cfg.Dag = dag
+	}
+
+	if introspectionJSONPath != "" {
+		introspectionJSON, err := os.ReadFile(introspectionJSONPath)
+		if err != nil {
+			return generator.Config{}, fmt.Errorf("read introspection json: %w", err)
+		}
+		cfg.IntrospectionJSON = string(introspectionJSON)
+	}
+
+	return cfg, nil
+}
```

Comment:
I think I can refactor that part to be command specific, correct me if I'm wrong there:
- generate-module always have a introspectionJSONPath set
- generate-client always have a introspectionJSONPath set and moduleID
- generate-library never has introspectionJSONPath

So technically, I could do that work per command and remove that `cfg.Close` function since I could close the client earlier (even prior to call `Generate`)

Reply:
Hm, I like that currently `generate-*` doesn't *assume* which one is set. I don't think we need it to be command specific (it's actually nice it's generic).

In this case, just because those are always true, we might change those in the future.